### DISCUSSION TITLE
Global versus loval dependencies

### DISCUSSION THREAD

**Discussion by Hubilari** [alire-project/alire#1407] [Support]
Hi,
trying to install openglada_glfw, as below :

$ alr get --build openglada_glfw
ⓘ Deploying openglada_glfw=0.9.0...                                      
##################################################################################################################################################### 100,0%
ⓘ Migrating lockfile from /home<rep.>/openglada_glfw_0.9.0_fc25165c/alire.lock to /home/<rep>/openglada_glfw_0.9.0_fc25165c/alire/alire.lock
ⓘ Deploying libglfw3=3.3.6...
The system package 'libglfw3-dev' is about to be installed.
This action might require admin privileges and impact your system installation.
Do you want Alire to install this system package?
[Y] Yes  [N] No  [A] Always  (default is Yes) 
ult is Yes) 

-I see that installation could impact my all system & not only the crate.
-I thought the logic of the crate was local, meaning dependencies could be different from a crate to another.
So it it really true?  If it is true, what's the difference with a global dependencies system?
If not why Alire asks this question, please?
Thanks
Mark

**Comment by mosteo** [comment]
Alire has a way to reuse system packages that aren't in Ada or that are otherwise overkill to redistribute as sources because they're generally available. We call those external dependencies. This is also a way to leverage very large/complex popular dependencies.

When one of those is needed, Alire can do for you the regular way to install, e.g., `sudo apt-get install whatever`. For some package managers, like `apt` in particular, the `--no-remove` flag will be used to ensure that the installation wouldn't cause a removal of packages, so it's quite safe to use. If you don't trust `alr` to do the right thing, you can always install these system packages in advance, and the next time `alr` will not try to install them and just use them.

So yes, these external dependencies are different in that they're not downloaded to the workspace, and to be used they need to be available in the system; but those are a way to simplify distribution of other crates, not the main purpose or way of working of `alr`.

**Comment by mosteo** [comment]
BTW, you can use `alr search --list --external` to see all crates in the index that are external. They're marked with `E` or `S`. The difference is the latter are a dependency of some other crate in the listing and have been detected as available in your system, while the former are know to exist but their detection hasn't been attempted.

Or you can use `alr search --list --external-detect`, and all of them will try to be detected, so all of them will be marked as either `S` (available with a known version) or `EU` (external unavailable).
